iphone/Classes/TiMediaAudioPlayerProxy.m

```diff
@@ -47,300 +44,501 @@ - (NSString *)apiName
+
+  // Buffering
+  [[[self player] currentItem] addObserver:self forKeyPath:@"playbackBufferEmpty" options:NSKeyValueObservingOptionNew context:nil];
+  [[[self player] currentItem] addObserver:self forKeyPath:@"playbackBufferFull" options:NSKeyValueObservingOptionNew context:nil];
+
+  // Timed metadata
+  [[[self player] currentItem] addObserver:self forKeyPath:@"timedMetadata" options:NSKeyValueObservingOptionNew context:nil];
 }
 
-#pragma mark Delegates
+- (void)removeNotificationObserver
+{
+  NSNotificationCenter *nc = [NSNotificationCenter defaultCenter];
+
+  if ([TiUtils isIOS10OrGreater]) {
+    [[self player] removeObserver:self forKeyPath:@"timeControlStatus"];
+  } else {
+    [[self player] removeObserver:self forKeyPath:@"rate"];
+  }
 
-- (void)playbackStateChanged:(id)sender
+  [nc removeObserver:self name:AVPlayerItemFailedToPlayToEndTimeNotification object:nil];
+  [nc removeObserver:self name:AVPlayerItemDidPlayToEndTimeNotification object:nil];
+
+  [[[self player] currentItem] removeObserver:self forKeyPath:@"playbackBufferEmpty"];
+  [[[self player] currentItem] removeObserver:self forKeyPath:@"playbackBufferFull"];
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey, id> *)change context:(void *)context
 {
-  if ([self _hasListeners:@"change"]) {
-    NSDictionary *event = [NSDictionary dictionaryWithObjectsAndKeys:[self state], @"state", [self stateToString:player.state], @"description", nil];
-    [self fireEvent:@"change" withObject:event];
+  if ([TiUtils isIOS10OrGreater]) {
+    if (object == _player && [keyPath isEqualToString:@"timeControlStatus"]) {
+      [self handleTimeControlStatusNotification:nil];
+    }
+  } else {
+    if (object == _player && [keyPath isEqualToString:@"rate"]) {
+      [self handlePlaybackStateChangeNotification:nil];
+    }
   }
-  if (player.errorCode != AS_NO_ERROR && player.state == AS_STOPPED) {
-    [[TiMediaAudioSession sharedSession] stopAudioSession];
+
+  if (object == _player.currentItem && [keyPath isEqualToString:@"playbackBufferEmpty"]) {
+    _state = TiAudioPlayerStateBuffering;
+  }
+
+  if (object == _player.currentItem && [keyPath isEqualToString:@"playbackBufferFull"]) {
+    _state = TiAudioPlayerStateWaitingForQueueToStart;
+  }
+
+  if (object == _player.currentItem && [keyPath isEqualToString:@"timedMetadata"]) {
+    [self handleTimedMetadataNotification:_player.currentItem];
+  }
+}
+
+// iOS < 10
+- (void)handlePlaybackStateChangeNotification:(NSNotification *)note
+{
+  TiAudioPlayerState oldState = _state;
+
+  switch (_player.status) {
+  case AVPlayerStatusUnknown:
+  case AVPlayerStatusFailed:
+    _state = TiAudioPlayerStateStopped;
+    break;
+  case AVPlayerStatusReadyToPlay:
+    if (_player.rate == 1.0) {
+      _state = TiAudioPlayerStatePlaying;
+    } else if (_player.currentItem.duration.value == _player.currentItem.currentTime.value || oldState == TiAudioPlayerStateStopping) {
```

Comment:
Please validate state with similar test case mentioned in ticket TIMOB-25801. I guess you have to adjust a bit.